crates/circuit/src/dag_node.rs

```diff
@@ -0,0 +1,329 @@
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use crate::circuit_instruction::CircuitInstruction;
+use pyo3::prelude::*;
+use pyo3::sync::GILOnceCell;
+use pyo3::types::{PyDict, PyList, PySequence, PyString, PyTuple};
+use pyo3::{intern, PyObject, PyResult};
+
+static SEMANTIC_EQ_PYTHON: GILOnceCell<PyResult<PyObject>> = GILOnceCell::new();
+
+/// Parent class for DAGOpNode, DAGInNode, and DAGOutNode.
+#[pyclass(module = "qiskit._accelerate.circuit", subclass)]
+#[derive(Clone, Debug)]
+pub struct DAGNode {
+    #[pyo3(get, set)]
+    pub _node_id: isize,
+}
+
+#[pymethods]
+impl DAGNode {
+    #[new]
+    #[pyo3(signature=(nid=-1))]
+    fn new(nid: isize) -> Self {
+        DAGNode { _node_id: nid }
+    }
+
+    fn __getstate__(&self) -> isize {
+        self._node_id
+    }
+
+    fn __setstate__(&mut self, nid: isize) {
+        self._node_id = nid;
+    }
+
+    fn __lt__(&self, other: &DAGNode) -> bool {
+        self._node_id < other._node_id
+    }
+
+    fn __gt__(&self, other: &DAGNode) -> bool {
+        self._node_id > other._node_id
+    }
+
+    fn __str__(_self: &Bound<DAGNode>) -> String {
+        format!("{}", _self.as_ptr() as usize)
+    }
+
+    fn __hash__(&self, py: Python) -> PyResult<isize> {
+        self._node_id.into_py(py).bind(py).hash()
+    }
+
+    /// Check if DAG nodes are considered equivalent, e.g., as a node_match for
+    /// :func:`rustworkx.is_isomorphic_node_match`.
+    ///
+    /// Args:
+    ///     node1 (DAGOpNode, DAGInNode, DAGOutNode): A node to compare.
+    ///     node2 (DAGOpNode, DAGInNode, DAGOutNode): The other node to compare.
+    ///     bit_indices1 (dict): Dictionary mapping Bit instances to their index
+    ///         within the circuit containing node1
+    ///     bit_indices2 (dict): Dictionary mapping Bit instances to their index
+    ///         within the circuit containing node2
+    ///
+    /// Return:
+    ///     Bool: If node1 == node2
+    #[staticmethod]
+    #[pyo3(signature = (*args), text_signature = "(node1, node2, bit_indices1, bit_indices2)")]
+    fn semantic_eq<'py>(py: Python<'py>, args: &Bound<PyTuple>) -> PyResult<Bound<'py, PyAny>> {
+        let semantic_eq_fn = SEMANTIC_EQ_PYTHON
+            .get_or_init(py, || {
+                let module = PyModule::import_bound(py, "qiskit.dagcircuit.dagnode")?;
+                Ok(module.getattr("_semantic_eq")?.unbind())
+            })
+            .as_ref()
+            .map_err(|e| e.clone_ref(py))?;
+        semantic_eq_fn.bind(py).call1(args)
+    }
+}
+
+/// Object to represent an Instruction at a node in the DAGCircuit.
+#[pyclass(module = "qiskit._accelerate.circuit", extends=DAGNode)]
+pub struct DAGOpNode {
+    pub instruction: CircuitInstruction,
+    #[pyo3(get)]
+    pub sort_key: PyObject,
+}
```

Comment:
What kind of interface do you have in mind for the future when these are true Rust-space classes and only accessed from within Rust?  We might be needing to access `_node_id` all the time, which is quite inconvenient to do from the Python-space forms of this.

(I'm assuming that this form of `DAGNode` doesn't really survive into Qiskit 2.0.)

Reply:
`DAGNode` and all of its subclasses provide an object-oriented model of working with a DAG, and this is inherently at odds with the design of `petgraph` (and largely Rust, IMO).

I'd think that in the future, we'd want to migrate to working with node IDs only for any graph manipulations, and require Rust callers to look up node weights given their IDs. So, we'd remove `_node_id` from the `DAGNode` struct altogether.